From 7d9e101603178d3ca8a83cfe8d0115a3fc700cfe Mon Sep 17 00:00:00 2001
From: Wojciech Dembinski <wojciech.dembinski@intel.com>
Date: Fri, 6 Aug 2021 10:07:54 +0200
Subject: [PATCH] Add node manager endpoint

It adds GET schema for the main node-manager node in Redfish.
Upstream-Status: Pending
---
 redfish-core/include/redfish.hpp         | 11 +++++++
 redfish-core/lib/managers.hpp            |  6 ++++
 static/redfish/v1/$metadata/index.xml    |  3 ++
 static/redfish/v1/schema/OemPower_v1.xml | 37 ++++++++++++++++++++++++
 4 files changed, 57 insertions(+)
 create mode 100644 static/redfish/v1/schema/OemPower_v1.xml

diff --git a/redfish-core/include/redfish.hpp b/redfish-core/include/redfish.hpp
index 54ba69d7..120c0f86 100644
--- a/redfish-core/include/redfish.hpp
+++ b/redfish-core/include/redfish.hpp
@@ -36,7 +36,12 @@
 #include "metric_report.hpp"
 #include "metric_report_definition.hpp"
 #include "network_protocol.hpp"
+#include "node-manager/domains_collection.hpp"
+#include "node-manager/node_manager.hpp"
+#include "node-manager/policies_collection.hpp"
 #include "node-manager/power.hpp"
+#include "node-manager/throttling_status.hpp"
+#include "node-manager/triggers.hpp"
 #include "pcie.hpp"
 #include "pcie_slots.hpp"
 #include "power_subsystem.hpp"
@@ -75,6 +80,12 @@ class RedfishService
      */
     explicit RedfishService(App& app)
     {
+        requestRoutesNodeManagerService(app);
+        requestRoutesNodeManagerDomains(app);
+        requestRoutesNodeManagerPolicies(app);
+        requestRoutesNodeManagerThrottlingStatus(app);
+        requestRoutesNodeManagerTriggers(app);
+
         requestAccountServiceRoutes(app);
 #ifdef BMCWEB_ENABLE_REDFISH_AGGREGATION
         requestRoutesAggregationService(app);
diff --git a/redfish-core/lib/managers.hpp b/redfish-core/lib/managers.hpp
index b50a3dc5..7fe7c343 100644
--- a/redfish-core/lib/managers.hpp
+++ b/redfish-core/lib/managers.hpp
@@ -1943,6 +1943,12 @@ inline void requestRoutesManager(App& app)
         // default oem data
         nlohmann::json& oem = asyncResp->res.jsonValue["Oem"];
         nlohmann::json& oemOpenbmc = oem["OpenBmc"];
+        nlohmann::json& oemIntel = oem["Intel"];
+        oemIntel["@odata.type"] = "#OemManager.Intel";
+        oemIntel["@odata.id"] = "/redfish/v1/Managers/bmc#/Oem/Intel";
+        oemIntel["NodeManager"] = {
+            {"@odata.id", "/redfish/v1/Managers/bmc/Oem/Intel/NodeManager"}};
+
         oem["@odata.type"] = "#OemManager.Oem";
         oem["@odata.id"] = "/redfish/v1/Managers/bmc#/Oem";
         oemOpenbmc["@odata.type"] = "#OemManager.OpenBmc";
diff --git a/static/redfish/v1/$metadata/index.xml b/static/redfish/v1/$metadata/index.xml
index 6fea9f86..846dcd98 100644
--- a/static/redfish/v1/$metadata/index.xml
+++ b/static/redfish/v1/$metadata/index.xml
@@ -3349,6 +3349,9 @@
     <edmx:Reference Uri="/redfish/v1/schema/OemComputerSystem_v1.xml">
         <edmx:Include Namespace="OemComputerSystem"/>
     </edmx:Reference>
+    <edmx:Reference Uri="/redfish/v1/schema/OemPower_v1.xml">
+        <edmx:Include Namespace="OemPower"/>
+    </edmx:Reference>
     <edmx:Reference Uri="/redfish/v1/schema/OemUpdateService_v1.xml">
         <edmx:Include Namespace="OemUpdateService"/>
     </edmx:Reference>
diff --git a/static/redfish/v1/schema/OemPower_v1.xml b/static/redfish/v1/schema/OemPower_v1.xml
new file mode 100644
index 00000000..58cf46aa
--- /dev/null
+++ b/static/redfish/v1/schema/OemPower_v1.xml
@@ -0,0 +1,37 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<edmx:Edmx xmlns:edmx="http://docs.oasis-open.org/odata/ns/edmx" Version="4.0">
+
+    <edmx:Reference Uri="http://docs.oasis-open.org/odata/odata/v4.0/errata03/csd01/complete/vocabularies/Org.OData.Core.V1.xml">
+        <edmx:Include Namespace="Org.OData.Core.V1" Alias="OData"/>
+    </edmx:Reference>
+    <edmx:Reference Uri="http://redfish.dmtf.org/schemas/v1/Resource_v1.xml">
+        <edmx:Include Namespace="Resource"/>
+        <edmx:Include Namespace="Resource.v1_0_0"/>
+    </edmx:Reference>
+
+    <edmx:DataServices>
+
+    <Schema xmlns="http://docs.oasis-open.org/odata/ns/edm" Namespace="OemPower">
+
+        <ComplexType Name="PowerControl" BaseType="Resource.OemObject">
+            <Annotation Term="OData.Description" String="OEM Extension for PowerControl."/>
+            <Annotation Term="OData.LongDescription" String="OEM Extension for PowerControl."/>
+            <Property Name="PowerLimitStorage" Type="OemPower.PowerLimitStorage" Nullable="true">
+                <Annotation Term="OData.Permissions" EnumMember="OData.Permission/ReadWrite"/>
+                <Annotation Term="OData.Description" String="Storage configuration of power limit."/>
+                <Annotation Term="OData.LongDescription" String="This property shall contain value defining configuration for power limit storage option."/>
+            </Property>
+        </ComplexType>
+
+        <EnumType Name="PowerLimitStorage">
+            <Member Name="Volatile">
+                <Annotation Term="OData.Description" String="Power limit is only stored in volatile memory."/>
+            </Member>
+            <Member Name="Persistent">
+                <Annotation Term="OData.Description" String="Power limit is stored in non-volatile memory."/>
+            </Member>
+        </EnumType>
+    </Schema>
+
+    </edmx:DataServices>
+</edmx:Edmx>
\ No newline at end of file
-- 
2.25.1

